### PR TITLE
[core] Rules with the same regex properties should be equal

### PR DESCRIPTION
**Pull request by adangel**
## Describe the PR

Found this while upgrading PMD Eclipse Plugin to 7.0.0. In the plugin, we often verify that we have the same ruleset active by executing `RuleSet#equals()`. This doesn't work as soon as a rule uses `java.util.regex.Pattern` based properties. The reason is, that `Pattern` doesn't override `equals` - see also https://stackoverflow.com/questions/52103427/why-these-two-java-util-pattern-are-not-equal

## Ready?

- [ ] Added unit tests for fixed bug/feature
- [ ] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [ ] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 6 violations,<br />
introduces 1 new violations, 0 new errors and 0 new configuration errors,<br />
removes 1 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/56ad80e7034843349b64c092850a615d/diff1/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/AbstractRule.java [hunk 1]
```diff
@@ -378,8 +381,32 @@ public boolean equals(Object o) {
...
+                that.getPropertiesByPropertyDescriptor()
+                        .forEach((key, value) -> {
+                            Object newValue = value;
+                            if (newValue instanceof Pattern) {
+                                newValue = ((Pattern) newValue).pattern();
+                            }
+                            thatPropertiesWithValues.put(key, newValue);
+                        });
+
+                equality = propertiesWithValues.equals(thatPropertiesWithValues);
```
Comment (jsotuyod) [inline pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/AbstractRule.java:408]: First of all, amazing catch!

I'd consider an alternative approach…

1. Have `PropertySource` expose a method for property equality.
2. Have an implementation of that in `AbstractPropertySource` that instead of relying on `Map.equals`, checks wether the keysets match, and delegates to the `PropertyDescriptor` the equality check between 2 values for that property.

I think that is going to be clearer, applies to all `PropertySource`s, and forces any future `PropertyDescriptor` to provide the equality check without us having to remember to come and change this.

Comment (oowekyala) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/AbstractRule.java:408]: I like your approach and agree that the problem is mostly that we had been using Map.equals. I think your new approach should be considered when implementing #3869. Instead of handling Maps of properties, we should be using `PropertyBundle` as a strongly typed container for properties and their values. Then we can just use `rule.getProperties().equals(otherRule.getProperties())` to implement equals, where `getProperties()` returns a property bundle, and equals is implemented like you describe.

I would prefer we reconsider that after #4380 is merged.

Comment (adangel) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/AbstractRule.java]: I've moved the implementation of equals now into AbstractPropertySource. Added some more tests for some edge cases (property descriptors, defaults, values).

I hope, this change doesn't make it more difficult to move the equals whereever it goes with #4380 (the comment there in PropertyDescriptor says, it should go to Rule::equals).
Btw. we only compare currently the name of a property descriptor and neither description nor default value. #4380 changes that to include the description.

**Review by oowekyala — APPROVED** [review]
